lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3384,6 +3385,17 @@ namespace {
           methods.push_back(MD);
           continue;
         }
+
+        if (auto dest = dyn_cast<DestructorDecl>(member)) {
+          // Only add this the destructor if it was provided by the user.
+          // Otherwise, this wont be a trivial type when it maybe should be.
+          if (!cast<clang::CXXDestructorDecl>(dest->getClangDecl())->isUserProvided())
```

Comment:
Why is it better to let Swift handle destruction of C++ types? Destroying a C++ type can be quite difficult in the general case (think of base classes, virtual base classes and other complexity that Swift does not have). I think it is better to let Clang destroy instances of C++ types unless we are certain that it is a trivial operation.

Reply:
Fair enough. I agree we should do that. When I went to change this I discovered an issue, though. For your `MyNonTrivialType` example, clang won't create a destructor for `MyNonTrivialType` because it expects that the struct will just be exploded and the destructor will be called on the individual members. When we try to call `MyNonTrivialType`'s destructor we get a linker error. [Here's a Godbolt example.](https://godbolt.org/z/mHn7k_)

Reply:
I suppose we could change this check, though. We could add all destructors and then in IRGen just check if the destructor is defined or not.

Reply:
If I dump the AST in your example, I see that Foo has a CXXDestructorDecl. Would Clang not IRGen it for us if we asked for it?

Reply:
There may be a way to get clang to do that, I'm not sure. Right now it will declare the destructor but, not define it.

Reply:
Can we use `Sema::DefineImplicitDestructor` to generate the body?

Reply:
Yes, I believe that's the way to do this. See the analogous case for implicit default constructors here:

https://github.com/apple/swift/pull/30630/files#diff-5c658ca134bff93fc4d435a34ecd5307R3509

Reply:
Great! I'll use that. I didn't realize we had access to clang sema (at the time).